src/libexpr/primops.cc

```diff
@@ -3712,16 +3712,29 @@ static RegisterPrimOp primop_toString({
 static void prim_substring(EvalState & state, const PosIdx pos, Value * * args, Value & v)
 {
     int start = state.forceInt(*args[0], pos, "while evaluating the first argument (the start offset) passed to builtins.substring");
-    int len = state.forceInt(*args[1], pos, "while evaluating the second argument (the substring length) passed to builtins.substring");
-    NixStringContext context;
-    auto s = state.coerceToString(pos, *args[2], context, "while evaluating the third argument (the string) passed to builtins.substring");
 
     if (start < 0)
         state.debugThrowLastTrace(EvalError({
             .msg = hintfmt("negative start position in 'substring'"),
             .errPos = state.positions[pos]
         }));
 
+
+    int len = state.forceInt(*args[1], pos, "while evaluating the second argument (the substring length) passed to builtins.substring");
+
+    // Special-case on empty substring to avoid O(n) strlen
+    // This allows for the use of empty substrings to efficently capture string context
+    if (len == 0) {
```

Comment:
```suggestion
    if (len <= 0) {
```

Reply:
If `len < 0` substring will take the remainder of the string. In particular, changing this behavior breaks at least `lib.strings.sanitizeDerivationName` and possibly other things in nixpkgs.

Reply:
:nauseated_face: 

That's not how negative numbers should work, but I guess we inherited that from C++.
If I ever order -1 pizza, please don't bury my neighbourhood under the world production of pizzas.

Reply:
I think the way we ended up here, is that len is an int, and substr expects a size_t, which should be unsigned. So on a 64 bit system, -1 is equivalent to 18446744073709551615. It's really unfortunate that this behavior is being used in practice. I'm not sure if there's a good way to deprecate this though.

Reply:
We can't really change the builtins, because that would affect reproducibility of old expressions, but we can improve Nixpkgs `lib`, which isn't subject to such a stability promise. It'd have to do a deprecation cycle to change the meaning of `substring` - emitting warnings for negative lengths before disallowing it. Then after some time it may adopt the cleaner semantics.
We could make the new `substring` just as fast as the old one by providing an extra builtin.

I doubt that this is worth doing though. Maybe a function `lib.strings.drop`, analogous to `lib.list.drop` might be nice. Solves part of the ugliness.